### PR TITLE
TMA-775: make attribute polling smart again

### PR DESCRIPTION
**Pull request by kubamahnert**

### CODE REVIEW

**Comment by kubamahnert** [comment]
extended test

File: lib/gooddata/models/metadata/label.rb [hunk 1]
```diff
@@ -40,19 +40,30 @@ def find_element_value(element_id)
...
+      end
+
+      results = client.post(final_url, 'validElementsRequest' => request_payload)
+
+      # Implementation of polling is based on
+      # https://opengrok.intgdc.com/source/xref/gdc-backend/src/test/java/com/gooddata/service/dao/ValidElementsDaoTest.java
+      status_url = results['uri']
+      if status_url
+        results = client.poll_on_response(status_url) do |body|
+          status = body['taskState'] && body['taskState']['status']
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Calls 'body['taskstate']' 2 times (DuplicateMethodCall)

File: lib/gooddata/models/metadata/label.rb [hunk 2]
```diff
@@ -40,19 +40,30 @@ def find_element_value(element_id)
...
-      # TMA-775 - the validElements API can possibly return more matches than requested (usually 1)
-      # so we do a preliminary first request to check and then increase the limit if needed
-      if results['validElements']['paging']['total'].to_i != params[:limit]
-        params[:limit] = 100_000
-        results, = valid_elements params
-        if params[:filter]
-          results['validElements']['items'] = results['validElements']['items'].reject do |i|
-            i['element']['title'] != params[:filter]
-          end
+    def get_valid_elements(url_or_params = {}, request_payload = {})
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Has approx 12 statements (TooManyStatements)

File: lib/gooddata/models/metadata/label.rb [hunk 3]
```diff
@@ -21,7 +21,7 @@ def find_value_uri(value)
       if items.empty?
         fail(AttributeElementNotFound, value)
       else
-        items.first['element']['uri']
+        items.find { |i| i['element']['title'] == value }['element']['uri']
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Has the variable name 'i' (UncommunicativeVariableName)

File: lib/gooddata/models/metadata/label.rb [hunk 4]
```diff
@@ -40,19 +40,30 @@ def find_element_value(element_id)
...
+    def get_valid_elements(url_or_params = {}, request_payload = {})
+      final_url = url_or_params
+
+      if url_or_params.is_a?(Hash)
+        default_params = {
+          limit: 50_000,
+          offset: 0,
+          order: 'asc'
+        }
+        params = default_params.merge(url_or_params).map { |x, v| "#{x}=#{CGI.escape(v.to_s)}" }.reduce { |acc, elem| "#{acc}&#{elem}" }
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Has the variable name 'v' (UncommunicativeVariableName)

Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Has the variable name 'x' (UncommunicativeVariableName)

File: lib/gooddata/models/user_filters/user_filter_builder.rb [hunk 5]
```diff
@@ -140,11 +140,9 @@ def self.verify_existing_users(filters, options = {})
 
     def self.create_label_cache(result, options = {})
       project = options[:project]
-      project_labels = project.labels
-
       result.reduce({}) do |a, e|
         e[:filters].map do |filter|
-          a[filter[:label]] = project_labels.find { |l| (l.identifier == filter[:label]) || (l.uri == filter[:label]) } unless a.key?(filter[:label])
+          a[filter[:label]] = project.labels(filter[:label]) unless a.key?(filter[:label])
```
Comment (yenkins) [inline lib/gooddata/models/user_filters/user_filter_builder.rb]: Calls 'filter[:label]' 3 times (DuplicateMethodCall)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/744/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/607/ : SUCCESS in 49s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/868/ : FAILURE in 35s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/746/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/608/ : SUCCESS in 42s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/869/ : SUCCESS in 39s

**Comment by kubamahnert** [comment]
extended test

**Comment by panjan** [comment]
ok to merge

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/747/ : SUCCESS in 14s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1209/ : FAILURE in 56m 34s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/968/ : SUCCESS in 41m 40s

**Comment by panjan** [comment]
@kubamahnert the build failure seems to be valid. Can you check that please?

**Pull request closed without merging**